core/src/main/java/com/linecorp/armeria/common/HttpRequestBuilder.java

```diff
@@ -0,0 +1,371 @@
+     *            .queryParams(QueryParams.of("from", "foo", "limit", 10))
+     *            .build(); // GET `/endpoint?from=foo&limit=10`
+     * }</pre>
+     * @see QueryParams
+     */
+    public HttpRequestBuilder queryParams(QueryParams queryParams) {
+        requireNonNull(queryParams, "queryParams");
+        final QueryParamsBuilder queryParamsBuilder = this.queryParams.toBuilder();
+        for (String name : queryParams.names()) {
+            queryParamsBuilder.add(name, requireNonNull(queryParams.get(name), "query"));
+        }
+        this.queryParams = queryParamsBuilder.build();
+        return this;
+    }
+
+    /**
+     * Sets a cookie for this request. For example:
+     * <pre>{@code
+     * HttpRequest.get("/")
+     *            .cookie(Cookie.of("cookie", "foo"))
+     *            .build();
+     * }</pre>
+     * @see Cookie
+     */
+    public HttpRequestBuilder cookie(Cookie cookie) {
+        requireNonNull(cookie, "cookie");
+        cookies.add(cookie);
+        return this;
+    }
+
+    /**
+     * Sets multiple cookies for this request. For example:
+     * <pre>{@code
+     * HttpRequest.get("/")
+     *            .cookies(Cookies.of(Cookie.of("cookie1", "foo"),
+     *                                Cookie.of("cookie2", "bar")))
+     *            .build();
+     * }</pre>
+     * @see Cookies
+     */
+    public HttpRequestBuilder cookies(Cookies cookies) {
+        requireNonNull(cookies, "cookies");
+        this.cookies.addAll(cookies);
+        return this;
+    }
+
+    /**
+     * Makes this request a streaming request.
+     */
+    public HttpRequestBuilder streaming() {
+        this.streaming = true;
+        return this;
+    }
+
+    /**
+     * Creates a new {@link HttpRequest}.
+     */
+    public HttpRequest build() {
+        if (!pathParams.isEmpty()) {
+            path = replacePathParams();
+        }
+        if (!queryParams.isEmpty()) {
+            path += "?" + queryParams.toQueryString();
+        }
+        final RequestHeaders requestHeaders = buildRequestHeader();
+        if (streaming) {
+            return new DefaultHttpRequest(requestHeaders);
+        }
+        if (content.length() == 0) {
+            content.close();
+            if (httpTrailers.isEmpty()) {
+                return new EmptyFixedHttpRequest(requestHeaders);
+            } else {
+                return new OneElementFixedHttpRequest(requestHeaders, httpTrailers);
+            }
+        }
+        if (httpTrailers.isEmpty()) {
+            return new OneElementFixedHttpRequest(requestHeaders, content);
+        } else {
+            return new TwoElementFixedHttpRequest(requestHeaders, content, httpTrailers);
+        }
+    }
+
+    private RequestHeaders buildRequestHeader() {
+        requireNonNull(path, "path");
+        requireNonNull(method, "method");
+        final RequestHeadersBuilder requestHeadersBuilder = RequestHeaders.builder().method(method).path(path);
```

Comment:
Perhaps you could keep the `RequestHeadersBuilder` as a member field of this builder and add headers and cookies directly to it?

Reply:
Sure, we can add headers directly to `RequestHeadersBuilder`, but cookies need to be set once in the end because there can be only one cookie header (right?)

Reply:
That's correct.